write_hooks.go

```diff
@@ -0,0 +1,223 @@
+// written to the content addressed filesystem
+type WriteHook func(ctx context.Context, f File, added map[string]string) (io.Reader, error)
+
+// hookFile configures a callback function to be executed on a saved
+// file, at a specific point in the merkelization process
+type hookFile struct {
+	// file for delayed hook calls
+	File
+	// once mutex for callback execution
+	once sync.Once
+	// slice of pre-merkelized paths that need to be saved before the hook
+	// can be called
+	requiredPaths []string
+	// function to call
+	callback WriteHook
+}
+
+// Assert hookFile implements HookFile at compile time
+var _ WriteHookFile = (*hookFile)(nil)
+
+// WriteHookFile is a file that can hook into the merkelization process, affecting
+// contents as contents are being rendered immutable
+type WriteHookFile interface {
+	File
+	RequiredPaths() []string
+	HasRequiredPaths(paths map[string]string) bool
+	CallAndAdd(ctx context.Context, adder Adder, added map[string]string) error
+}
+
+// NewWriteHookFile wraps a File with a hook & set of sibling / child dependencies
+func NewWriteHookFile(file File, cb WriteHook, requiredPaths ...string) File {
+	if file.IsDirectory() {
+		panic("cannot create a WriteHookFile with a Directory")
+	}
+
+	return &hookFile{
+		File:          file,
+		requiredPaths: requiredPaths,
+		callback:      cb,
+	}
+}
+
+func (h *hookFile) RequiredPaths() []string {
+	return h.requiredPaths
+}
+
+func (h *hookFile) HasRequiredPaths(merkelizedPaths map[string]string) bool {
+	for _, p := range h.requiredPaths {
+		if _, ok := merkelizedPaths[p]; !ok {
+			log.Debugf("hook %q can't fire. waiting for %s", h.FullPath(), p)
+			return false
+		}
+	}
+	return true
+}
+
+func (h *hookFile) CallAndAdd(ctx context.Context, adder Adder, merkelizedPaths map[string]string) (err error) {
+	h.once.Do(func() {
+		log.Debugf("calling hook path=%s merkelized=%#v", h.FullPath(), merkelizedPaths)
+		var r io.Reader
+		r, err = h.callback(ctx, h.File, merkelizedPaths)
+		if err != nil {
+			return
+		}
+		if err = adder.AddFile(ctx, NewMemfileReader(h.FullPath(), r)); err != nil {
+			return
+		}
+	})
+
+	return err
+}
+
+// WriteWithHooks writes a file or directory to a given filestore using
+// merkelization hooks
+// failed writes are rolled back with delete requests for all added files
+func WriteWithHooks(ctx context.Context, fs Filesystem, root File) (string, error) {
+	var (
+		finalPath       string
+		waitingHooks    []WriteHookFile
+		doneCh          = make(chan error, 0)
+		addedCh         = make(chan AddedFile, 1)
+		merkelizedPaths = map[string]string{}
+	)
+
+	addFS, ok := fs.(AddingFS)
+	if !ok {
+		return "", ErrNotAddingFS
+	}
+
+	adder, err := addFS.NewAdder(ctx, true, true)
+	if err != nil {
+		return "", err
+	}
+
+	var rollback = func() {
+		log.Debug("rolling back failed write operation")
+		// TODO (b5) - we can't run rollbacks without a "soft delete" command that
```

Comment:
If I understand this correctly, let's go through an example. Say file1 has "A,B,C" and writes them all. Then file2 has "A,D,E" and attempts to write them. Suppose "A" succeeds and "D" succeeds but "E" fails. The rollback should delete "D" but *not* "A" because it's in use by file1. Is that right?

Isn't this equivalent to running gc over the store? If so, I don't really think it's something this package should even attempt. Just let the orphan blocks live in the store, and allow external processes to clean up when they get around to it.

Reply:
your example is exactly right, and you're right all this would be covered by GC when running with IPFS. Not all FS's work this way though, so either implement the same reference tracking & garbage collection across all stores, or emit delete requests that garbage-collected filesystems can ignore. Not sure which is right quite yet